Fix/address page state navigation 2

ygshbht opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/23725
PROPOSAL: https://github.com/Expensify/App/issues/23725#issuecomment-1656792082


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

**Verify for places where you have to select an American State from a list**:

For **Personal Address**:

1. Go to Profile > Address > State (with US as country)
2. Click back button of device
3. Notice it Properly navigates to Home address screen

For **debit card page**:

1. Go to /settings/wallet/add-debit-card or /send/new/add-debit-card > State
2. Click back button of device
3. Notice it properly navigates to Debit card screen

For **ReimbusementAccountPage**

1. Select a Workspace > Settings > Bank Account
2. Verify for all State Selectors you see in the flow (e.g Company Address, Incorporation State, etc.)

- 1.  Click back button of device when in respective StateSelectionPages
- 2.  Notice it properly navigates to previous screen


- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as tests
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/65986357/b4deae99-c426-444f-94b5-45602eb87847


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/65986357/383ec463-d4a5-4827-8454-5b3934c52e4f


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/65986357/c7aa615d-13cb-4a35-b195-741d3556f6ba


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/65986357/93c9e5de-a097-4358-a6f5-a6b91d3ab71c


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/65986357/b54f8816-59eb-492a-950e-d6cf99565032


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/65986357/55b3fcc3-5cae-4292-84b9-023dae47e1db


<!-- add screenshots or videos here -->

</details>

ygshbht committed:
Convert StatePicker modal to StateSelection page

ygshbht committed:
Incorporation state resets bugfix

ygshbht committed:
Modify pages to accomodate for StateSelection

ygshbht committed:
Merge https://github.com/Expensify/App into fix/address-page-state-navigation

ygshbht committed:
Refactor to use useGeographicalStateFromRoute hook

ygshbht committed:
move getStateFromRoute to useGeographicalStateFromRoute

ygshbht committed:
Add displayname to StateSelectorWithRef

ygshbht committed:
Merge https://github.com/Expensify/App into fix/address-page-state-navigation

ygshbht committed:
Fix AddressPage: State not changing on autocomplete

ygshbht committed:
Refactor StateSelector logic

ygshbht committed:
Refactor AddressPage

ygshbht committed:
linting fixes StateSelector

ygshbht committed:
linting

ygshbht committed:
StateSelector and related files refactoring

ygshbht committed:
Merge https://github.com/Expensify/App into fix/address-page-state-navigation

ygshbht committed:
Fix incorrect merge

ygshbht committed:
Reverted scripts/shellCheck.sh to state at 33f15e8d85b3366b1a09bca914f61d5caa4a2fb5

ygshbht committed:
Linting

ygshbht committed:
Merge https://github.com/Expensify/App into fix/address-page-state-navigation

ygshbht committed:
Comment formatting src/hooks/useGeographicalStateFromRoute.ts

Co-authored-by: Rajat Parashar <parasharrajat@users.noreply.github.com>

ygshbht committed:
Rename useStateFomrUrl to shouldUseStateFromUrl

ygshbht committed:
forwardedRef type correction StateSelector

ygshbht committed:
StateSelector: rename paramName to queryParam

ygshbht committed:
Remove lodashGet from useGeographicalStateFromRoute

ygshbht committed:
StateSelectionPage appendParam restructuring

ygshbht committed:
Prettier formatting

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented:
Old [PR](https://github.com/Expensify/App/pull/29053)

ygshbht commented on src/components/StateSelector.tsx:
```diff
@@ -0,0 +1,100 @@
+                description={label || translate('common.state')}
+                onPress={() => {
+                    const activeRoute = Navigation.getActiveRoute();
+                    // @ts-expect-error Navigation.navigate does take a param
```
I had to add `// @ts-expect-error Navigation.navigate does take a param` at places because TS was giving me error. But i dont think this problem is specific to my code. What do you say?

mountiny replied:
cc @blazejkustra @fabioh8010 for visibility

fabioh8010 replied:
@ygshbht I've checked here and it's specific to your code. It's happening for two reasons:

1. You are making `SETTINGS_ADDRESS_STATE`'s `getRoute` function return a `string` [here](https://github.com/Expensify/App/blob/77f236c7822570bf9f31db1d2563c852033291fa/src/ROUTES.ts#L136), but [we have a condition](https://github.com/Expensify/App/blob/77f236c7822570bf9f31db1d2563c852033291fa/src/ROUTES.ts#L554) to make the `Route` type be `never` if we have a route that returns `string`. This is to ensure all routes return only constants, I think you can solve it with this code:
    ```ts
        SETTINGS_ADDRESS_STATE: {
            route: 'settings/profile/address/state',
            getRoute: (state: string, backTo?: string, label?: string) =>
                `${getUrlWithBackToParam(`settings/profile/address/state?state=${state}`, backTo)}${label ? (`&label=${encodeURIComponent(label)}` as const) : ''}` as const,
        },
    ```
    The adjusted code will make `getRoute` return all the possible route strings as constant values instead of just `string`, thus eliminating the problem.
2. After doing the first fix, if you remove the `@ts-expect-error` comment you will have another error:
    <img width="1010" alt="Screenshot 2024-02-20 at 09 42 28" src="https://github.com/Expensify/App/assets/20051562/ced15857-a096-456a-8d9d-e5735abf53a9">
    Since `stateCode` can be `undefined` and your route function doesn't accept that, to fix you can do `stateCode ?? ''` if it's safe enough, or change `getRoute` signature to accept `undefined` values there and handle it, or put a condition to only call the navigation if `stateCode` is defined.
    ```ts
                    onPress={() => {
                        if (!stateCode) {
                            return;
                        }
    
                        const activeRoute = Navigation.getActiveRoute();
                        Navigation.navigate(ROUTES.SETTINGS_ADDRESS_STATE.getRoute(stateCode, activeRoute, label));
                    }}
    ```
    The best solution will depends on how this fix/feature works.

ygshbht replied:
Thanks, I've made the change

mountiny commented on src/pages/settings/Profile/PersonalDetails/AddressPage.js:
```diff
@@ -63,7 +65,12 @@ function AddressPage({privatePersonalDetails, route}) {
     const {translate} = useLocalize();
     const address = useMemo(() => lodashGet(privatePersonalDetails, 'address') || {}, [privatePersonalDetails]);
-    const countryFromUrl = lodashGet(route, 'params.country');
+
+    const countryFromUrlTemp = lodashGet(route, 'params.country');
+    // check if country is valid
+    const countryFromUrl = lodashGet(CONST.ALL_COUNTRIES, countryFromUrlTemp) ? countryFromUrlTemp : '';
+    const stateFromUrl = useGeographicalStateFromRoute();
```
Why exactly do we need this check for `countryFromUrlTemp`
```suggestion
    const countryFromUrlTemp = lodashGet(route, 'params.country');

    // Check if country is valid
    const countryFromUrl = lodashGet(CONST.ALL_COUNTRIES, countryFromUrlTemp) ? countryFromUrlTemp : '';
    const stateFromUrl = useGeographicalStateFromRoute();
```

mountiny commented on src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx:
```diff
@@ -0,0 +1,121 @@
+                Navigation.goBack(appendParam(backTo, 'state', option.value));
+            } else {
+                // Otherwise, navigate to the specific route defined in "backTo" with a country parameter
+                // @ts-expect-error Navigation.navigate does take a paraml
```
```suggestion
                // @ts-expect-error Navigation.navigate does take a param
```

mountiny reviewed (commented):
@ygshbht is this ready for a review/ testing?

ygshbht commented:
@mountiny Yes, it is

ygshbht committed:
Comment typo fix src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Comment formatting src/pages/settings/Profile/PersonalDetails/AddressPage.js

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

mountiny commented:
Seems like some conflicts just came in, I will trigger a build once you resolve them, thank you!

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented:
@mountiny Resolved

ygshbht committed:
Fix TS errors

mountiny commented:
@parasharrajat ready for your review

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented on src/ROUTES.ts:
```diff
@@ -131,6 +131,16 @@ const ROUTES = {
         route: 'settings/profile/address/country',
         getRoute: (country: string, backTo?: string) => getUrlWithBackToParam(`settings/profile/address/country?country=${country}`, backTo),
     },
+    SETTINGS_ADDRESS_STATE: {
+        route: 'settings/profile/address/state',
```
What do you propose? `CountrySelector `still has similar path
![image](https://github.com/Expensify/App/assets/65986357/66ec043b-7278-4006-bc99-81d0fab4f209)

mountiny replied:
I think we can keep it same as because the country selector is not updated either

ygshbht committed:
Refactor StateSelectionPage with more explanatory comments

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented:
Merged @parasharrajat

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

parasharrajat reviewed (changes requested):
BUG ⬆️

ygshbht commented:
Thanks @parasharrajat. Before i set up a new remote Mac instance to work on this, could you please help me with the following? 
1. Do you see this issue with the existing website's StateSelectorModal, CountrySelectionPage, or other components?
2. Is the header space less or more? Reference to a place where the space is proper

ygshbht commented on src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx:
```diff
@@ -0,0 +1,129 @@
+        <ScreenWrapper
+            testID={StateSelectionPage.displayName}
+            includeSafeAreaPaddingBottom={false}
+            style={[styles.pb0]}
+            includePaddingTop={false}
```
These settings are taken straight out of `StateSelectorModal`.
https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/components/StatePicker/StateSelectorModal.tsx#L80-L81

But, now that i see, I dont find these props in other pages (not modals) like `LegalNamePage `, `AddressPage`, `CountrySelectionPage`. 
https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/pages/settings/Profile/PersonalDetails/AddressPage.tsx#L98-L101

https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/pages/settings/Profile/PersonalDetails/LegalNamePage.tsx#L80-L84

https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/pages/settings/Profile/DisplayNamePage.js#L78-L82

https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/pages/settings/Profile/PersonalDetails/CountrySelectionPage.tsx#L62-L65
Apologies for the oversight.

ygshbht committed:
Incorrect header space fix: src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx

Co-authored-by: Rajat <parasharrajat@users.noreply.github.com>

ygshbht committed:
Remove unused code StateSelectionPage.tsx

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented on src/pages/settings/Profile/PersonalDetails/AddressPage.tsx:
```diff
@@ -45,19 +45,17 @@ function AddressPage({privatePersonalDetails, route}: AddressPageProps) {
-    const address = useMemo(() => lodashGet(privatePersonalDetails, 'address') || {}, [privatePersonalDetails]);
-    const countryFromUrlTemp = lodashGet(route, 'params.country');
-
+    const address = useMemo(() => privatePersonalDetails?.address, [privatePersonalDetails]);
```
Is this part of this PR? Should it be? 
https://github.com/Expensify/App/blob/48657b48505e40fecc7c6f43da903aadb12d549a/src/pages/settings/Profile/PersonalDetails/AddressPage.tsx#L46

parasharrajat reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
Prettier formatting

mountiny commented:
I was ooo sorry. @ygshbht can you merge main to resolve the TS failure?

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

mountiny commented on src/hooks/useGeographicalStateFromRoute.ts:
```diff
@@ -0,0 +1,16 @@
+type CustomParamList = ParamListBase & Record<string, Record<string, string>>;
+type State = keyof typeof COMMON_CONST.STATES;
+
+export default function useGeographicalStateFromRoute(stateParamName = 'state'): State | undefined {
```
Could you add a docs/ comment to this hook to explain in plain english what it does, also ideally give an example

mountiny commented on src/components/StateSelector.tsx:
```diff
@@ -0,0 +1,99 @@
+type State = keyof typeof COMMON_CONST.STATES;
+
+type StateSelectorProps = {
+    /** Form error text. e.g when no country is selected */
```
```suggestion
    /** Form error text. e.g when no state is selected */
```

mountiny commented on src/libs/Url.ts:
```diff
@@ -41,4 +41,18 @@ function hasSameExpensifyOrigin(url1: string, url2: string): boolean {
+ * Appends or updates a query parameter in a given URL.
+ */
+function appendParam(url: string, paramName: string, paramValue: string) {
+    if (url.includes(`${paramName}=`)) {
+        // If parameter exists, replace it
```
```suggestion
    // If parameter exists, replace it
    if (url.includes(`${paramName}=`)) {
```

mountiny commented on src/libs/Url.ts:
```diff
@@ -41,4 +41,18 @@ function hasSameExpensifyOrigin(url1: string, url2: string): boolean {
+        const regex = new RegExp(`${paramName}=([^&]*)`);
+        return url.replace(regex, `${paramName}=${paramValue}`);
+    }
+    // If parameter doesn't exist, append it
```
```suggestion

    // If parameter doesn't exist, append it
```

mountiny commented on src/pages/settings/Profile/PersonalDetails/AddressPage.tsx:
```diff
@@ -46,7 +48,10 @@ function AddressPage({privatePersonalDetails, route}: AddressPageProps) {
     const {translate} = useLocalize();
     const address = useMemo(() => privatePersonalDetails?.address, [privatePersonalDetails]);
-    const countryFromUrl = route.params?.country;
+    const countryFromUrlTemp = route?.params?.country;
```
```suggestion

    const countryFromUrlTemp = route?.params?.country;
```

mountiny commented on src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx:
```diff
@@ -0,0 +1,125 @@
+    const {translate} = useLocalize();
+
+    const [searchValue, setSearchValue] = useState('');
+
+    const params = route.params as RouteParams | undefined; // Type assertion
+
+    const currentState = params?.state;
+    const label = params?.label;
```
```suggestion
    const params = route.params as RouteParams | undefined;
    const currentState = params?.state;
    const label = params?.label;
```

mountiny replied:
To simplify this

mountiny commented:
Reviewed @ygshbht typecheck still seems to be failing

ygshbht committed:
fix: handle potential 'undefined' object issue

ygshbht committed:
Typo fix src/components/StateSelector.tsx

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Format src/libs/Url.ts

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Format src/libs/Url.ts

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Format src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Format src/pages/settings/Profile/PersonalDetails/AddressPage.tsx

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

mountiny commented on src/pages/settings/Profile/PersonalDetails/AddressPage.tsx:
```diff
@@ -46,7 +48,11 @@ function AddressPage({privatePersonalDetails, route}: AddressPageProps) {
     const {translate} = useLocalize();
     const address = useMemo(() => privatePersonalDetails?.address, [privatePersonalDetails]);
-    const countryFromUrl = route.params?.country;
+
+    const countryFromUrlTemp = route?.params?.country;
+    // Check if country is valid
```
Sorry this was my bad, the suggestion should be like this
```suggestion
    const countryFromUrlTemp = route?.params?.country;

    // Check if country is valid
```

mountiny commented on src/ROUTES.ts:
```diff
@@ -131,6 +131,16 @@ const ROUTES = {
+
+        getRoute: (state?: string, backTo?: string, label?: string) =>
+            `${getUrlWithBackToParam(`settings/profile/address/state${state ? `?state=${encodeURIComponent(state)}` : ''}`, backTo)}${
+                // Nullish operator ?? doesnt seem to be a replacement for || here
```
```suggestion
                // the label param can be an empty string so we cannot use a nullish ?? operator
```

mountiny commented:
@ygshbht before merging, can you please update the tests to cover every page where the state picker is used? 

@parasharrajat can you please test them all on web and one narrow platform?

The total reward has already been increased to $1500 so I think we need to make sure to properly test this PR. Regressions would lead to penalties which we do not want

ygshbht committed:
Comment for function useGeographicalStateFromRoute

mountiny commented on src/hooks/useGeographicalStateFromRoute.ts:
```diff
@@ -5,6 +5,10 @@ import {CONST as COMMON_CONST} from 'expensify-common/lib/CONST';
 
+/**
+ * Extracts the 'state' (default) query parameter from the route/ url and validates it against COMMON_CONST.STATES, returning its ISO code or `undefined`.
+ * Example: const stateISO = useGeographicalStateFromRoute(); // Assuming 'state' param is 'CA' or another valid state, returns the corresponding ISO code or `undefined` if invalid.
```
The example I meant more so an example of the the URL and what it returns than how to use it in code

ygshbht committed:
hash append for StateSelector edge cases

ygshbht commented:
@parasharrajat I found this issue for which i pushed https://github.com/Expensify/App/pull/36770/commits/3ba97a550a07ce38adbf5e397bba3e56907552b1 commit. Do you think that it is a very edge case and the code complexity is unnecessary for that, in which case i can revert the commit?

https://github.com/Expensify/App/assets/65986357/a0552f10-96e2-48dd-8dbe-c454bcf15214

ygshbht committed:
Comment update useGeographicalStateFromRoute

ygshbht committed:
Add StateSelector to MoneyRequest Navigation stack to work for /send/new/add-debit-card

ygshbht commented:
@parasharrajat I also happened to find that though `StateSelector` worked for Debit card when it is in this route `/settings/wallet/add-debit-card`, it didnt work when it is in ` /send/new/add-debit-card`. I made changes to Stack Navigation files to fix this in commit https://github.com/Expensify/App/pull/36770/commits/57bcee217ff3a01861f0a0f8ac6a09bba7a4bfc3

ygshbht commented:
@mountiny I have updated the tests

ygshbht committed:
Formatting src/pages/settings/Profile/PersonalDetails/AddressPage.tsx

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Commnet udpate src/ROUTES.ts

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Comment update ROUTES.ts

mountiny commented on src/libs/Navigation/linkingConfig/config.ts:
```diff
@@ -486,6 +490,7 @@ const config: LinkingOptions<RootStackParamList>['config'] = {
                         [SCREENS.IOU_SEND.ENABLE_PAYMENTS]: ROUTES.IOU_SEND_ENABLE_PAYMENTS,
                         [SCREENS.IOU_SEND.ADD_BANK_ACCOUNT]: ROUTES.IOU_SEND_ADD_BANK_ACCOUNT,
                         [SCREENS.IOU_SEND.ADD_DEBIT_CARD]: ROUTES.IOU_SEND_ADD_DEBIT_CARD,
+                        [SCREENS.MONEY_REQUEST.STATE_SELECTOR]: {path: ROUTES.MONEY_REQUEST_STATE_SELECTOR.route, exact: true},
```
Could you move this up to the other money request routes?

mountiny commented on src/hooks/useGeographicalStateFromRoute.ts:
```diff
@@ -0,0 +1,28 @@
+
+/**
+ * Extracts the 'state' (default) query parameter from the route/ url and validates it against COMMON_CONST.STATES, returning its ISO code or `undefined`.
+ * Example 1: Url: https://dev.new.expensify.com:8082/settings/profile/address?state=MO Returns: MO
+ * Example 2: Url: https://dev.new.expensify.com:8082/settings/profile/address?state=ASDF Returns: undefined
+ * Example 3: Url: https://dev.new.expensify.com:8082/settings/profile/address Returns: undefined
+ * Example 4: Url: https://dev.new.expensify.com:8082/settings/profile/address?state=MO-hash-a12341 Returns: MO
```
```suggestion
 * Example 1: Url: https://new.expensify.com/settings/profile/address?state=MO Returns: MO
 * Example 2: Url: https://new.expensify.com/settings/profile/address?state=ASDF Returns: undefined
 * Example 3: Url: https://new.expensify.com/settings/profile/address Returns: undefined
 * Example 4: Url: https://new.expensify.com/settings/profile/address?state=MO-hash-a12341 Returns: MO
```

ygshbht committed:
Example update src/hooks/useGeographicalStateFromRoute.ts

Co-authored-by: Vit Horacek <36083550+mountiny@users.noreply.github.com>

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
Refactor linking config StateSelector

mountiny commented:
@parasharrajat ready for your retest

ygshbht commented on src/ROUTES.ts:
```diff
@@ -420,6 +430,16 @@ const ROUTES = {
         getRoute: (iouType: ValueOf<typeof CONST.IOU.TYPE>, transactionID: string, reportID: string) => `create/${iouType}/start/${transactionID}/${reportID}/scan` as const,
     },
 
+    MONEY_REQUEST_STATE_SELECTOR: {
+        route: 'moneyrequest/state',
+
+        getRoute: (state?: string, backTo?: string, label?: string) =>
+            `${getUrlWithBackToParam(`moneyrequest/state${state ? `?state=${encodeURIComponent(state)}` : ''}`, backTo)}${
+                // the label param can be an empty string so we cannot use a nullish ?? operator
+                // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+                label ? `${backTo || state ? '&' : '?'}label=${encodeURIComponent(label)}` : ''
+            }` as const,
+    },
```
No. i think its not because of the url but because it needs to be in the same `StackNavigator `, otherwise when doing `Navaigation.goBack()`, it is not able to modify the url and just goes back to the previous page without adding `state` param

https://github.com/Expensify/App/blob/9ad7ca0ad76fda24c04b42cc352482a2554b7f8b/src/libs/Navigation/AppNavigator/ModalStackNavigators.tsx#L80

https://github.com/Expensify/App/blob/9ad7ca0ad76fda24c04b42cc352482a2554b7f8b/src/libs/Navigation/AppNavigator/ModalStackNavigators.tsx#L204

ygshbht commented:
> > it didnt work when it is in  /send/new/add-debit-card
> 
> What are steps to open this page? @ygshbht

@parasharrajat  Sorry, I'm not really sure. I bumped into this url in the repository when looking for places where `StateSelector `is used. Currently i'm accessing this page directly via URL. If i try to traceback by searching for this url and `IOU_SEND_ADD_DEBIT_CARD` in repository, it suggests it might be used in `Settlement `button 
![image](https://github.com/Expensify/App/assets/65986357/e0ac1720-af3b-40f8-82c0-a989891c6003)

Maybe here, but the button doesnt work for me 
![image](https://github.com/Expensify/App/assets/65986357/b96858e1-69e1-43dc-b4d7-53720b4a02cd)

ygshbht committed:
Revert "hash append for StateSelector edge cases"

This reverts commit 3ba97a550a07ce38adbf5e397bba3e56907552b1.

ygshbht committed:
Sync URL 'state' param with StateSelector value

ygshbht commented on src/hooks/useGeographicalStateFromRoute.ts:
```diff
@@ -0,0 +1,28 @@
+/**
+ *  See {@link module:src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx#withHash} for more information.
+ */
+const removeHash = (arg: string): string => arg.replace(/-hash-.*$/, '');
```
Yes, that was my initial thought but I thought that would add an extra url in browser's history. Now that i see, that's not the case. So i have removed the old commit and implemented your change

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
change moneyrequest/state url to request/state

Co-authored-by: Rajat <parasharrajat@users.noreply.github.com>

ygshbht committed:
change moneyrequest/state url to request/state

Co-authored-by: Rajat <parasharrajat@users.noreply.github.com>

ygshbht commented:
@parasharrajat I suppose thats just how `react-navigation `works. I see some similar issues online https://github.com/react-navigation/react-navigation/issues/10046

https://github.com/Expensify/App/assets/65986357/58691d34-49fa-4ec2-9727-906e2e50ed33

ygshbht commented:
@parasharrajat This issue is already there in the production website with other selectors such as `CurrencySelctor ` and `CountrySelector`

https://github.com/Expensify/App/assets/65986357/728e5981-287a-445b-9862-e8f1a786cae5

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
StateSelector bugfix - for issue causing url to not update in browser url bar

ygshbht commented:
@parasharrajat This seems to be stemming from the code syncing url `state `value with the browser url value. I have modified it to remove the `state `from completely from url, aligning more with how the `CountrySelector `does 

https://github.com/Expensify/App/blob/0fa4d0f2a200d58af452fa6c37bdc27b7f7fe2d5/src/components/AddressForm.tsx#L153

ygshbht commented:
@parasharrajat  I see a few other minor issues, let me get back to you shortly

ygshbht commented:
@parasharrajat This is one issue i encounterd - when you try to use `StateSelector` after changing country. Is it something do you think is due to issues in code of this PR or an edge case revealed due to issues in the `SelectionList `component? The error goes away if i dont pass the `shouldUseDynamicMaxToRenderPerBatch` prop 

![image](https://github.com/Expensify/App/assets/65986357/604270c2-4068-4210-b787-97abf9c70dbb)


https://github.com/Expensify/App/assets/65986357/7f0caebc-9eb7-4791-910d-605a6ef72b22

ygshbht commented:
`shouldUseDynamicMaxToRenderPerBatch` prop was introduced in this commit https://github.com/Expensify/App/commit/6c876c4dee0c858009d156f2e56c27b00a0ac422. Though I dont see this issue in existing `StateSelectorModal` in prod, would you prefer me to remove this prop as it seems to solve an seemingly insignificant issue

ygshbht commented:
@parasharrajat No, I dont mean to say that its not present in prod. 
So to go about fixing it, do you think there should be a new issue as its probably internal to `SelectionList`?

ygshbht committed:
useGeographicalStateFromRoute - check for state before accessing stateISO

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
Fix: Inconsistent validation error for required field in home address

ygshbht commented:
@parasharrajat When you used `StateSelector`, and didnt select any country, it didnt show any error, which i have fixed through this https://github.com/Expensify/App/pull/36770/commits/2619a29d6ef297a1c357aadfa9dc0ad06cd68272 commit following the same logic as `CountrySelector`. I have also refactored the code in `StateSelector` by removing the local state (react state) for the state (geographical state)

ygshbht committed:
Remove unused code: StateSelector

ygshbht commented:
@parasharrajat, I have been studying various parts of the repository to address the problems and merge this long-standing issue as soon as possible. As demonstrated in the video, a similar page, `ProfilePage`, which also utilizes `SelectionList`, exhibits the same issue when I make those unrelated edits. From my observation, `SelectionList` attempts to match the scroll position of the previous `SelectionList`. Thus, in instances where you select `United States`, which is located towards the bottom of the `CountrySelector` list of 150 items (not using the search input, as it shortens the list), and then attempt to use `StateSelector` (with about 50 states), the list scrolls to the bottom, showing the last states (or directly errors out if `shouldUseDynamicMaxToRenderPerBatch` is also used). Therefore, if you mimic the `ProfilePage` solution by simply adding an empty `View`, it resolves our problem without introducing side effects.


https://github.com/Expensify/App/assets/65986357/3b953dce-165c-4b2a-8375-edb3f7f0137b

ygshbht committed:
StateSelectionPage: Fix issue of page crash when country is changed multiple times

ygshbht commented:
@parasharrajat I have pushed the patch in https://github.com/Expensify/App/pull/36770/commits/4efccc0d43af2788a24b767165a483bb21ea213e. 

https://github.com/Expensify/App/assets/65986357/d761e24a-da8c-4fe7-9c46-4c2d6050cfb3

ygshbht committed:
es lint comment formatting

ygshbht committed:
eslint Formatting fix

ygshbht committed:
prettier

mountiny commented:
Sorry I was ooo, I think that the latest comment suggests @ygshbht found a solution to this issue. @parasharrajat can you please test if that is working as expected tomorrow? thank you!

mountiny commented:
@ygshbht Can you please resolve conflicts?

@parasharrajat can you conclude the testing please?

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht commented on src/components/StateSelector.tsx:
```diff
@@ -0,0 +1,103 @@
+    const didOpenStateSelector = useRef(false);
+    const isFocused = useIsFocused();
+
+    useEffect(() => {
+        if (isFocused && didOpenStateSelector.current) {
+            didOpenStateSelector.current = false;
+            if (!stateFromUrl) {
+                onBlur?.();
+            }
+        }
+
+        if (!stateFromUrl) {
+            return;
+        }
+
+        // This will cause the form to revalidate and remove any error related to country name
+        if (onInputChange) {
+            onInputChange(stateFromUrl);
+        }
+
+        Navigation.setParams({state: undefined});
+
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [stateFromUrl, onBlur, isFocused]);
+
+    const title = stateCode && Object.keys(COMMON_CONST.STATES).includes(stateCode) ? translate(`allStates.${stateCode}.stateName`) : '';
```
Yes, it won't work if not used within form. But I suppose since it in an input component, it will always be used inside a form. Do you want it to be more independent, generic, futureproof, and have its own state?

ygshbht replied:
Currently it is not used without a `Form `and If it is used as a standalone component, it can always be wrapped around a `Form `like `IncorporationStateBusiness`. Or the devs can modify the code as per their needs. For current uses and its intended purpose, do you think it still needs a local state?
![image](https://github.com/Expensify/App/assets/65986357/2880a200-b24e-4b9e-90cd-90ed0c42bab6)

ygshbht replied:
Current [StatePicker](https://github.com/Expensify/App/blob/main/src/components/StatePicker/index.tsx) too doesnt have local state. Since we're just mimicking its functionality, do you still think we need to worry about local state?

ygshbht replied:
Done

ygshbht committed:
Merge branch 'main' into fix/address-page-state-navigation-2

ygshbht committed:
Add explanatory comments to StateSelector

parasharrajat reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

mountiny commented on src/components/StateSelector.tsx:
```diff
@@ -0,0 +1,107 @@
+    /** Callback to call when the picker modal is dismissed */
+    onBlur?: () => void;
+
+    /** object to get route details from */
```
NAB
```suggestion
    /** Object to get route details from */
```

mountiny commented on src/pages/settings/Profile/PersonalDetails/StateSelectionPage.tsx:
```diff
@@ -0,0 +1,126 @@
+            if (navigation.getState()?.routes.length === 1) {
+                // If this is the only page in the navigation stack (examples include direct navigation to this page via URL or page reload).
+                if (_.isEmpty(backTo)) {
+                    // No "backTo": default back navigation.
```
NAB: I think some of these comments might not have been needed
```suggestion
```

mountiny reviewed (approved):
Been a long journey, I wont hold this for longer, thank you very much for pushing this ahead

mountiny commented:
Thank you for patience and good work @ygshbht @parasharrajat 🙌

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8598730206) to production by https://github.com/Julesssss in version: 1.4.60-13 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
